lib/logstash/inputs/beats.rb

```diff
@@ -0,0 +1,173 @@
+
+  # TODO(sissel): Add CA to authenticate clients with.
+  BUFFERED_QUEUE_SIZE = 1
+  RECONNECT_BACKOFF_SLEEP = 0.5
+  
+  def register
+    require "lumberjack/server"
+    require "concurrent"
+    require "logstash/circuit_breaker"
+    require "logstash/sized_queue_timeout"
+
+    if !@ssl
+      @logger.warn("SSL Certificate will not be used") unless @ssl_certificate.nil?
+      @logger.warn("SSL Key will not be used") unless @ssl_key.nil?
+    elsif !ssl_configured?
+      raise LogStash::ConfigurationError, "Certificate or Certificate Key not configured"
+    end
+
+    @logger.info("Starting Beats input listener", :address => "#{@host}:#{@port}")
+    @lumberjack = Lumberjack::Server.new(:address => @host, :port => @port,
+      :ssl => @ssl, :ssl_certificate => @ssl_certificate, :ssl_key => @ssl_key,
+      :ssl_key_passphrase => @ssl_key_passphrase)
+
+    # Create a reusable threadpool, we do not limit the number of connections
+    # to the input, the circuit breaker with the timeout should take care 
+    # of `blocked` threads and prevent logstash to go oom.
+    @threadpool = Concurrent::CachedThreadPool.new(:idletime => 15)
+
+    # in 1.5 the main SizeQueue doesnt have the concept of timeout
+    # We are using a small plugin buffer to move events to the internal queue
+    @buffered_queue = LogStash::SizedQueueTimeout.new(BUFFERED_QUEUE_SIZE)
+
+    @circuit_breaker = LogStash::CircuitBreaker.new("Beats input",
+                            :exceptions => [LogStash::SizedQueueTimeout::TimeoutError])
+
+  end # def register
+
+  def ssl_configured?
+    !(@ssl_certificate.nil? || @ssl_key.nil?)
+  end
+
+  def target_codec_on_field?
+    !@target_codec_on_field.empty?
+  end
+
+  def run(output_queue)
+    start_buffer_broker(output_queue)
+
+    while !stop? do
+      # Wrapping the accept call into a CircuitBreaker
+      if @circuit_breaker.closed?
+        connection = @lumberjack.accept # call that creates a new connection
+        next if connection.nil? # if the connection is nil the connection was close.
+
+        invoke(connection, @codec.clone) do |event|
+          if stop?
+            connection.close
+            break
+          end
+
+          begin
+            @circuit_breaker.execute {
+              @buffered_queue.push(event, @congestion_threshold)
+            }
+          rescue => e
+            raise e
+          end
+        end
+      else
+        @logger.warn("Beats input: the pipeline is blocked, temporary refusing new connection.")
+        sleep(RECONNECT_BACKOFF_SLEEP)
+      end
+    end
+  end # def run
+
+  public
+  def stop
+    @lumberjack.close
+  end
+
+  private
+  def create_event(codec, map)
+    # Filebeats uses the `message` key and LSF `line`
+    message = map.delete("message")
+
+    @codec.decode(message) do |decoded|
+      decorate(decoded)
+      map.each { |k, v| decoded[k] = v; v.force_encoding(Encoding::UTF_8) }
```

Comment:
v might not be a string. can be a bool, a number, an array or another object. Does the json decoder ensure we're dealing with utf8 always?